utils/result_try.hh

```diff
@@ -0,0 +1,442 @@
+#pragma once
+
+#include <type_traits>
+#include "utils/result.hh"
+
+namespace utils {
+
+namespace internal {
+
+// Dummy result satisfying ExceptionContainerResult, for static asserts
+template<typename T = void>
+using dummy_result = result_with_exception<T, std::exception>;
+
+struct noop_converter {
+    template<typename T>
+    using wrapped_type = T;
+
+    template<typename T>
+    static auto wrap(T&& t) {
+        return std::move(t);
+    }
+
+    template<typename F, typename... Args>
+    static auto invoke(F&& f, Args&&... args) {
+        return f(std::forward<Args>(args)...);
+    }
+};
+
+struct futurizing_converter {
+    template<typename T>
+    using wrapped_type = typename seastar::futurize<T>::type;
+
+    template<typename T>
+    static auto wrap(T&& t) {
+        if constexpr (seastar::is_future<T>::value) {
+            return t;
+        } else {
+            return seastar::make_ready_future<T>(std::forward<T>(t));
+        }
+    }
+
+    template<typename F, typename... Args>
+    static auto invoke(F&& f, Args&&... args) {
+        return seastar::futurize_invoke(std::forward<F>(f), std::forward<Args>(args)...);
+    }
+};
+
+// TODO: Should we make a virtual base class for this?
+// This would allow using non-generic lambdas in result_catch and result_catch_dots.
+template<typename Handle, typename R>
+concept ExceptionHandle = ExceptionContainerResult<R> && requires (Handle handle) {
+    { handle.into_result() } -> std::same_as<R>;
+    { handle.into_future() } -> std::same_as<seastar::future<R>>;
+};
+
+template<ExceptionContainerResult R>
+struct failed_result_handle {
+private:
+    R&& _failed_result;
+
+public:
+    failed_result_handle(R&& r) : _failed_result(std::move(r)) {}
+
+    // Returns the failed result. Can be used only once as this moves out
+    // the failed result. This should be called once during the handler's
+    // lifetime, preferably as `return handler.into_failure()`.
+    R into_result() {
+        return std::move(_failed_result);
+    }
+
+    seastar::future<R> into_future() {
+        return seastar::make_ready_future<R>(std::move(_failed_result));
+    }
+};
+
+static_assert(ExceptionHandle<failed_result_handle<dummy_result<>>, dummy_result<>>);
+
+template<ExceptionContainerResult R>
+struct exception_ptr_handle {
+private:
+    std::exception_ptr _eptr;
+
+public:
+    exception_ptr_handle(std::exception_ptr&& eptr) : _eptr(std::move(eptr)) {}
+
+    // Throws the exception.
+    [[noreturn]] R into_result() {
+        std::rethrow_exception(std::move(_eptr));
+    }
+
+    seastar::future<R> into_future() {
+        return seastar::make_exception_future<R>(std::move(_eptr));
+    }
+};
+
+// Check that it is a valid handle with some arbitrary ExceptionContainerResult
+static_assert(ExceptionHandle<exception_ptr_handle<dummy_result<>>, dummy_result<>>);
+
+template<typename Ex, typename Cb>
```

Comment:
I don't think it will work. For example, some handler could explicitly set its return type to `exceptions::coordinator_result<>` and return the failed result via `return handle.into_result()`. If you try to check it with `failed_result_handle<dummy_result<>>` as argument, then you will get a type mismatch on the return statement.

Reply:
I think that the best we can do is to constrain everything in `result_try` and `result_futurize_try`, where you know all the necessary types. I plan to write those constraints.